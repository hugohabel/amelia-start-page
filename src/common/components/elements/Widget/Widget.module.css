/* Main Styles */
.mainContainer {
  background-color: var(--main-widget-bg-color);
  border-radius: .428rem;
  display: flex;
  margin: 0 24px 24px 0;
  padding: 16px;
  position: relative;
  width: 344px;
  color: var(--main-widget-text-color);
}
/* End Main Styles */

/* Actions */
.actionRemove {
  top: 8px;
  cursor: pointer;
  position: absolute;
  right: 12px;
  width: 12px;
}
.actionEdit {
  top: 3px;
  cursor: pointer;
  position: absolute;
  right: 12px;
  width: 12px;
}
/* End Actions */