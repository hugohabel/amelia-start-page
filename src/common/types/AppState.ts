import { CountdownFormat } from './amelia';

// App State
export interface IAppState {
  widgets: any[];
  sidebar: ISidebar;
}

// Sidebar
export interface ISidebar {
  isOpen: boolean;
  activeMenu: TActiveMenu;
  activeView: TActiveView;
}

// Countdown Widget
export interface ICountdownWidget {
  type: 'countdown';
  enabled: boolean;
  data: {
    events: IEvent[];
  };
}

export interface IEvent {
  id: string;
  name: string;
  date: string;
  format: CountdownFormat;
}
// End Countdown Widget

// Weather Widget
export interface IWeatherWidget {
  type: 'weather';
  enabled: boolean;
  data: {
    locations: Location[];
  };
}

export interface ILocation {
  id: string;
  lat: string;
  lng: string;
}
// End Weather Widget

// Auxiliary Types
export type TActiveMenu = 'add' | 'about' | 'config' | null;
export type TActiveView = 'add_countdown' | 'add_weather' | null;

// App State Actions
type TAction =
  | { type: 'addNewWidget'; widget: ICountdownWidget | IWeatherWidget }
  | { type: 'toggleSidebar' }
  | { type: 'setActiveMenu'; activeMenu: TActiveMenu }
  | { type: 'setActiveView'; activeView: TActiveView }
  | { type: 'addNewCountdownEvent'; event: IEvent }
  | { type: 'addNewWeatherLocation'; location: Location }
  | { type: 'removeWidgetItem'; uuid: string; widgetType: string };

export type TAppAction = TAction;
export type TAppDispatch = (action: TAppAction) => void;
