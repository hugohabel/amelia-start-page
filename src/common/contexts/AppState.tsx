// External Dependencies
import { createContext, ReactNode, useContext, useReducer } from 'react';
import { debounce } from 'lodash';

// Internal Dependencies
import { LOCAL_STORAGE_STATE_KEY } from '../constants';
import { TAppAction, TAppDispatch, IAppState } from '../types/AppState';
import { SafeLocalStorage } from '../utils/localStorage/localStorage';
import { initialAppState, rootReducer } from '../state/state';

// Interfaces + Types
interface IAppStateProviderProps {
  children: ReactNode;
  initialStateOverride?: IAppState;
}

// Utility Functions

/**
 * Retrieves state from local storage or initializes with default state.
 *
 * @param defaultState AppState   Default app state.
 * @returns AppState  State (can be an existing one, or a default one)
 */
function getStateFromLocalStorageOrInitialize(defaultState: IAppState) {
  const stateFromLocalStorage = SafeLocalStorage.getInstance().getItem(LOCAL_STORAGE_STATE_KEY);
  if (stateFromLocalStorage) {
    return JSON.parse(stateFromLocalStorage) as IAppState;
  }
  return defaultState;
}

/**
 * Persists app state in memory or in local storage
 *
 * @param state AppState  State to be persisted.
 */
function persistState(state: IAppState) {
  SafeLocalStorage.getInstance().setItem(LOCAL_STORAGE_STATE_KEY, JSON.stringify(state));
}

const debouncePersistState = debounce(persistState, 500, {
  leading: false,
  trailing: true,
  maxWait: 1000,
});

function reducerWrapper(state: IAppState, action: TAppAction) {
  const nextState = rootReducer(state, action);
  debouncePersistState(nextState);
  return nextState;
}

// Create Contexts
const AppStateContext = createContext<IAppState>(initialAppState);
const AppDispatchContext = createContext<TAppDispatch>(() =>
  // eslint-disable-next-line no-console
  console.warn('Calling default dispatch is a no-op!')
);

/**
 * App State Provider
 *
 * @param param AppStateProviderProps Props to initialize the app state provider.
 * @returns JSX App Dispatch and App State Contexts.
 */
export function AppStateProvider({ children, initialStateOverride }: IAppStateProviderProps) {
  const hydratedInitialAppState = {
    ...initialAppState,
  };

  const [state, dispatch] = useReducer(
    reducerWrapper,
    initialStateOverride ?? hydratedInitialAppState,
    getStateFromLocalStorageOrInitialize
  );

  return (
    <AppDispatchContext.Provider value={dispatch}>
      <AppStateContext.Provider value={state}>{children}</AppStateContext.Provider>
    </AppDispatchContext.Provider>
  );
}

AppStateProvider.defaultProps = {
  initialStateOverride: undefined,
};

export const useAppState = () => useContext(AppStateContext);
export const useAppDispatch = () => useContext(AppDispatchContext);
