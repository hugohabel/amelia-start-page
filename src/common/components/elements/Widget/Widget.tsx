// External Dependencies
import EditAttributesOutlinedIcon from '@material-ui/icons/EditAttributesOutlined';

// Internal Dependencies
import { CloseIcon } from '../../icons/CloseIcon';
import { useAppDispatch } from '../../../contexts/AppState';

// Styles
import styles from './Widget.module.css';

// Types
interface IWidgetProps {
  // eslint-disable-next-line no-undef
  children: JSX.Element;
  style?: any;
  uuid: string;
  widgetType: string;
  editAction: boolean;
}

function Widget({ children, style, uuid, widgetType, editAction }: IWidgetProps) {
  const dispatch = useAppDispatch();

  // Handlers
  // eslint-disable-next-line no-shadow
  function handleRemove(uuid: string, widgetType: string) {
    dispatch({
      type: 'removeWidgetItem',
      uuid,
      widgetType,
    });
  }

  return (
    <>
      {/* Main Container */}
      <div style={style} className={styles.mainContainer}>
        {/* Inherited Content */}
        {children}

        {/* Widget Actions */}
        {editAction && (
          <div className={styles.actionEdit}>
            <EditAttributesOutlinedIcon fontSize="small" />
          </div>
        )}
        <div
          className={styles.actionRemove}
          onClick={() => {
            handleRemove(uuid, widgetType);
          }}
          data-testid={uuid}>
          <CloseIcon />
        </div>
        {/* End Widget Actions */}
      </div>
      {/* End Main Container */}
    </>
  );
}

export { Widget };
