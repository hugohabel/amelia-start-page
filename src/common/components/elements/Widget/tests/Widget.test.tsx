// External Dependencies
import { fireEvent, getByTestId, render, waitFor } from '@testing-library/react';

// Internal Dependencies
import { AppStateProvider, useAppDispatch } from '../../../../../common/contexts/AppState';
import { Widget } from '../Widget';

// Mocked Functions
jest.mock('../../../../../common/contexts/AppState', () => ({
  ...(jest.requireActual('../../../../../common/contexts/AppState') as object),
  useAppDispatch: jest.fn(),
}));
let mockAppDispatch: ReturnType<typeof useAppDispatch>;

// Mocked Values
const UUID = 'a2c82ff0-a886-4016-a9ac-e3b6d8e49676';

describe('Widget', () => {
  it('matches snapshot', async () => {
    const component = render(
      <Widget style={{backgroundColor: "#FFFFFF"}} uuid={UUID} widgetType="test" editAction={false}>
        <h1>Widget Container - Children</h1>
      </Widget>
    );

    expect(component.asFragment()).toMatchSnapshot();
  });

  describe('Widget - Handlers', () => {

    beforeEach(() => {
      mockAppDispatch = jest.fn();
      (useAppDispatch as jest.Mock).mockReturnValue(mockAppDispatch);
    });

    it('triggers remove handler when clicking on the remove icon', async () => {
      const component = render(
        <AppStateProvider>
          <Widget style={{backgroundColor: "#FFFFFF"}} uuid={UUID} widgetType="test" editAction={false}>
            <h1>Widget Container - Children</h1>
          </Widget>
        </AppStateProvider>
      );

      const removeButtonElement = await component.findByTestId(UUID);
      expect(removeButtonElement).toBeInTheDocument();

      fireEvent.click(getByTestId(component.container, UUID));

      await waitFor(() => {
        expect(mockAppDispatch).toHaveBeenCalledWith({
          type: 'removeWidgetItem', 
          uuid: 'a2c82ff0-a886-4016-a9ac-e3b6d8e49676',
          widgetType: 'test'
        });
      });
    });
  });
});