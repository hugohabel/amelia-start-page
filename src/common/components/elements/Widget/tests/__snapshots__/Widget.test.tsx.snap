// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widget matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mainContainer"
    style="background-color: rgb(255, 255, 255);"
  >
    <h1>
      Widget Container - Children
    </h1>
    <div
      class="actionRemove"
      data-testid="a2c82ff0-a886-4016-a9ac-e3b6d8e49676"
    >
      <svg
        fill="none"
        height="16"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="18"
          x2="6"
          y1="6"
          y2="18"
        />
        <line
          x1="6"
          x2="18"
          y1="6"
          y2="18"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
